keps/sig-storage/2639-secret-protection/README.md

```diff
@@ -0,0 +1,450 @@
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a feature to protect secrets while it is in use. Currently, user can delete a Secret that is being used by other resources, like Pods, PVs, and VolumeSnapshots. This may have negative impact on the resouces using the Secret and it may result in data loss.
+
+Similar features for protecting PV and PVC already exist as [pv-protection](https://github.com/kubernetes/enhancements/issues/499) and [pvc-protection](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md).
+
+
+## Motivation
+
+This feature aims to protect secrets from being deleted while they are in-use.
+Secrets can be used by below ways:
+- From Pod:
+  - [Mounted as data volumes](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod)
+  - [Exposed as environment variables](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables)
+  - [Generic ephemeral volumes
+](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes) (can be handled as CSI PV below)
+- From PV:
+  - [CSI](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html):
+    - provisioner secret
+    - controller publish secret
+    - node stage secret
+    - node publish secret
+    - controller expand secret
+  - non-CSI:
+    - dependent on each storage driver and will be deprecated soon (Out of scope)
+- [From VolumeSnapshot](https://kubernetes-csi.github.io/docs/secrets-and-credentials-volume-snapshot-class.html):
+  - snapshotter secret
+
+### Goals
+
+- Protect secrets from being deleted while they are in use.
+
+### Non-Goals
+
+- Protect important secrets that aren't in use from being deleted
+- Protect resources other than secrets from being deleted.
```

Comment:
why not configmaps? don't the same issues affect those?

Reply:
The main purpose of this proposal is to avoid failure in deleting volume. The example use case is referenced from [here](https://github.com/kubernetes/kubernetes/issues/101130). Failure in deleting the backend volume will make resource leak, and loosing a way to access the volume is a kind of data corruption (This proposal is trying to protect PV and VolumeSnapshot, not secret itself, in this sense). Configmap is not used in such a way and won't have the same level of impact, so current focus is only secret.

I agree that Configmap will have the similar issue to the used-by-pod use case of this proposal. It would be nice to have, but it is worth considering to drop from the scope depending on the implementation cost, which should be the purpose of non-goal in KEP (Of course, if there are big impact by configmap deletion and it can be implemented easily, I will add it to the scope).

Reply:
As you might guess, the background of my comment above is that the KEP merge deadline for 1.22 is today. I personally feel that Configmap should be protected too, but I just would like to provide a feature to protect secret, ASAP.
(I hope that we can change the scope later if needed, instead of being forced KEP process for almost the same feature).